### PR TITLE
🚑 Fix FormData encoding (#1741)

### LINKED ISSUE #1741: FormData.fromMap is encoding request body automatically
### Package

dio

### Version

5.0.2

### Output of `flutter doctor -v`

_No response_

### Dart Version

2.19.2

### Steps to Reproduce

```
Future<FormData> mapToFormData(
  Map<String, dynamic> formData,
) async {
  try {
    return FormData.fromMap(formData);
  } catch (e) {
    rethrow;
  }
}
```

### Expected Result

The above code should return request body as:

- item[0][product_size_id]: 14

### Actual Result

Instead of the expected result the code is returning request body as:

- items%5B0%5D%5Bproduct_size_id%5D: 14

### PR DESCRIPTION
**Pull request by kuhnroyal**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### Additional context and info (if any)
Fixes #1741

### CODE REVIEW

File: dio/lib/src/utils.dart [hunk 1]
```diff
@@ -38,8 +38,8 @@ String encodeMap(
 }) {
   final urlData = StringBuffer('');
   bool first = true;
-  final leftBracket = isQuery ? '[' : '%5B';
-  final rightBracket = isQuery ? ']' : '%5D';
+  final leftBracket = isQuery || !encode ? '[' : '%5B';
```
Comment (AlexV525) [inline dio/lib/src/utils.dart:41]: I think we can add some related comments to avoid breaking again.

**Review by AlexV525 — APPROVED** [review]
LGTM with a comment suggestion

**Comment by AlexV525** [comment]
After merging this, you can comment with a new version in the release comment to publish the fix ASAP.

File: dio/lib/src/utils.dart (hunk 1 above)
Comment (kuhnroyal) [reply dio/lib/src/utils.dart:44]: Done

**Pull request merged**